Add annotation search/verification score filtering

hudson-newey opened the pull request:
# Add annotation score filtering

## Changes

- Adds "Minimum Score" and "Maximum Score" inputs to the annotation search / verification parameters form & associated `score` qsp. This qsp has a lower and upper bound (similar to date/time qsps). E.g. `&score=0.2,0.8`
- Changes the `debounced-input.component` into a directive that supports two-way `[(value)]` bindings
- Fixes bug where clicking on a typeahead input label would select the first typeahead on the page, not the associated typeahead.
- Typeahead labels are now slightly closer (0.25rem) to their input fields
- Standardizes tooltip hints to a dotted underline
- Remove unused `typeInElement` html helper. I previously replaced this helper method with the ng-neat spectator [`typeInElement`](https://github.com/ngneat/spectator/blob/05c09ed0e4c26b62e9e84f88c86d909ad69d1683/projects/spectator/src/lib/type-in-element.ts#L11) method
- Hard lock ng-neat spectator version because new version broke TypeScript types for component factory class inference
- Improves typing for `filterAnd` helper so that it can infer the underlying AbstractModel type without explicitly specifying it in the generic type arguments
- Array query string parameters with all null / undefined values will now be omitted from url parameters (e.g. `&score=,` will not be emitted)

## Issues

Fixes: #2317

## Problems (for future issues)

- Min/max inputs for score should be a slider with two inputs

## Visual Changes

![image](https://github.com/user-attachments/assets/42079952-e602-4a62-8db6-4950205cb0a6)

_New annotation search page with "Minimum Score" and "Maximum Score" filters_

Additionally, notice that the labels above the typeaheads are slightly closer to the typeahead input, making it easier to distinguish what input the label applies to.

---

![image](https://github.com/user-attachments/assets/a12166d1-3a72-4aa4-8e5a-b6c883faa335)

_Verification parameter modal with new "score" input_

---

![image](https://github.com/user-attachments/assets/7e3c6186-974b-48ee-aa27-6235de83ddbd)

_Example or errors being shown for the score inputs_

---

![image](https://github.com/user-attachments/assets/e4fca37e-97e3-4728-a493-06dd41966322)

_Annotation import page now has a dotted underline instead of a solid underline. Indicating that there is an action on hover. This makes this hint consistent with the rest of the client._

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Backfill score filtering

hudson-newey committed:
Replace debounce component with debounce directive

hudson-newey committed:
percentage input and double binding debounce input

hudson-newey committed:
Working score filters

hudson-newey committed:
Change score to unbounded number value

hudson-newey committed:
Passing tests

hudson-newey committed:
Add search form tests

hudson-newey committed:
Add tests for new functionality

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.ts:
```diff
@@ -228,4 +234,13 @@ export class AnnotationSearchFormComponent implements OnInit {
+    }
+
+    return value;
+  }
```
seems general enough it should be elsewhere in our solution?

atruskie commented on src/app/directives/debouncedInput/debounced-input.directive.spec.ts:
```diff
@@ -0,0 +1,108 @@
+  });
+});
+
+// import { createHostFactory, SpectatorHost } from "@ngneat/spectator";
```
i assume these tests are to be deleted or redesigned?

atruskie commented on src/app/styles/_hints.scss:
love it

atruskie reviewed (approved):
I know it's not in the issue, but can we add a maximum score input as well please

hudson-newey committed:
Add toNumber helper

This function is much more complicated than I'd like to admit.

hudson-newey committed:
Structural typing for toNumber helper

TypeScript can now statically determine the output at compile /
typechecking time.

Example

```ts
toNumber("5") === toNumber("5.1")
	 ^ 5		   ^ 5.1
```

hudson-newey committed:
Add newline condition toNumber

hudson-newey committed:
Migrate debounced input directive tests

hudson-newey committed:
Upper score input and form validation

hudson-newey committed:
Add tests for ranged score parsing

hudson-newey committed:
Empty arrays are omitted in query string parameters

hudson-newey committed:
Add tests and pre-filling maximum score input

hudson-newey committed:
Handle null lower and upper bound cases

hudson-newey committed:
Hard lock ng-neat spectator version

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.html:
```diff
@@ -56,24 +56,61 @@
+          <span
+            class="input-group-text tooltip-hint"
+            [ngbTooltip]="
+              'Applying this filter will cause all events without a score to be hidden.'
```
```suggestion
              'Any event with an empty score, will also be filtered out.'
```

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.html:
```diff
@@ -56,24 +56,61 @@
+          <span
+            class="input-group-text tooltip-hint"
+            [ngbTooltip]="
+              'Applying this filter will cause all events without a score to be hidden.'
```
```suggestion
              'Any event with an empty score, will also be filtered out.'
```

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.ts:
```diff
@@ -231,16 +277,24 @@ export class AnnotationSearchFormComponent implements OnInit {
+    ) {
+      this.scoreRangeError.set("Score lower bound must be less than or equal to the score upper bound.");
+    } else {
+      this.scoreRangeError.set(null);
```
angular has built in validation for forms, why are we rolling our own?

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.ts:
```diff
@@ -231,16 +277,24 @@ export class AnnotationSearchFormComponent implements OnInit {
+      this.searchParameters.scoreUpperBound !== null &&
+      this.searchParameters.scoreLowerBound > this.searchParameters.scoreUpperBound
+    ) {
+      this.scoreRangeError.set("Score lower bound must be less than or equal to the score upper bound.");
```
```suggestion
      this.scoreRangeError.set("Score minimum must be less than or equal to the score maximum.");
```

Should use consistent language with the labels on the inputs

atruskie commented on src/app/helpers/query-string-parameters/queryStringParameters.spec.ts:
```diff
@@ -127,6 +127,44 @@ describe("queryStringParameters", () => {
+      expect(result).toEqual(expectedOutput);
+    });
+
+    it("should not keys that have empty array values", () => {
```
missing word in test title?

atruskie commented on src/app/helpers/typing/toNumber.ts:
great!

atruskie reviewed (approved):
I'd strongly encourage standard angular forms validation

hudson-newey committed:
Migrate to Angular form validation

hudson-newey committed:
Disable test that only fails in testing environment

I will re-enable this test later, but in the interest of time, I'm not
going to block this PR (and subsequently other peoples work) because I
can't get a test to work correctly (that I have manually verified
doesn't fail in prod / dev environment).

hudson-newey committed:
Fix linting errors
